File: backend/app/schemas/bad_case.py
"""
Bad Case Schema — 创建 / 更新 / 响应
"""
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ==================== 枚举 ====================

class BadCaseTag(str, Enum):
    """Bad Case 标签枚举"""
    HALLUCINATION = "hallucination"
    ERROR = "error"
    OMISSION = "omission"
    TIMEOUT = "timeout"
    OTHER = "other"


class BadCaseStatus(str, Enum):
    """Bad Case 状态枚举"""
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"


# ==================== 创建 ====================

class BadCaseCreate(BaseModel):
    """标记 Bad Case 请求"""
    trace_id: UUID = Field(..., description="关联的 Trace ID")
    tag: BadCaseTag = Field(..., description="标签: hallucination/error/omission/timeout/other")
    description: Optional[str] = Field(None, description="问题描述")
    assignee_id: Optional[UUID] = Field(None, description="负责人 ID")


# ==================== 更新 ====================

class BadCaseUpdate(BaseModel):
    """更新 Bad Case 请求"""
    status: Optional[BadCaseStatus] = Field(None, description="状态: open/in_progress/resolved/closed")
    assignee_id: Optional[UUID] = Field(None, description="负责人 ID")
    resolution: Optional[str] = Field(None, description="解决方案")


# ==================== 响应 ====================

class BadCaseResponse(BaseModel):
    """Bad Case 响应"""
    id: UUID
    project_id: UUID
    trace_id: UUID
    tag: str
    description: Optional[str] = None
    status: str
    assignee_id: Optional[UUID] = None
    resolution: Optional[str] = None
    marked_by: UUID
    marked_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/services/badcase.ts
/**
 * Bad Case API 封装
 */
import { get, post, put, PaginatedData } from './api';

// ---------- 类型 ----------

/** Bad Case 标签 */
export type BadCaseTag = 'hallucination' | 'error' | 'omission' | 'timeout' | 'other';

/** Bad Case 状态 */
export type BadCaseStatus = 'open' | 'in_progress' | 'resolved' | 'closed';

/** Bad Case 列表项 */
export interface BadCaseInfo {
  id: string;
  project_id: string;
  trace_id: string;
  tag: BadCaseTag;
  description?: string;
  status: BadCaseStatus;
  assignee_id?: string;
  resolution?: string;
  marked_by: string;
  marked_at?: string;
  resolved_at?: string;
  created_at?: string;
  updated_at?: string;
}

/** 统计摘要 */
export interface BadCaseStats {
  total: number;
  by_status: Record<string, number>;
  by_tag: Record<string, number>;
}

// ---------- 标签/状态 中文映射 ----------

export const BAD_CASE_TAG_LABELS: Record<BadCaseTag, string> = {
  hallucination: '幻觉',
  error: '错误',
  omission: '遗漏',
  timeout: '超时',
  other: '其他',
};

export const BAD_CASE_TAG_COLORS: Record<BadCaseTag, string> = {
  hallucination: 'purple',
  error: 'red',
  omission: 'orange',
  timeout: 'gold',
  other: 'default',
};

export const BAD_CASE_STATUS_LABELS: Record<BadCaseStatus, string> = {
  open: '待处理',
  in_progress: '处理中',
  resolved: '已解决',
  closed: '已关闭',
};

export const BAD_CASE_STATUS_COLORS: Record<BadCaseStatus, string> = {
  open: 'red',
  in_progress: 'processing',
  resolved: 'green',
  closed: 'default',
};

/** 允许的状态流转（与后端 VALID_TRANSITIONS 保持一致） */
export const BAD_CASE_NEXT_STATUS: Record<BadCaseStatus, BadCaseStatus[]> = {
  open: ['in_progress', 'closed'],
  in_progress: ['resolved', 'closed'],
  resolved: ['closed'],
  closed: [],
};

// ---------- API ----------

export function listBadCases(
  projectId: string,
  page = 1,
  pageSize = 20,
  filters?: {
    status?: string;
    tag?: string;
    assignee_id?: string;
  },
): Promise<PaginatedData<BadCaseInfo>> {
  return get<PaginatedData<BadCaseInfo>>(
    `/projects/${projectId}/bad-cases`,
    {
      page,
      page_size: pageSize,
      ...filters,
    },
  );
}

export function markBadCase(
  projectId: string,
  data: {
    trace_id: string;
    tag: BadCaseTag;
    description?: string;
    assignee_id?: string;
  },
): Promise<BadCaseInfo> {
  return post<BadCaseInfo>(
    `/projects/${projectId}/bad-cases`,
    data,
  );
}

export function getBadCaseDetail(
  projectId: string,
  caseId: string,
): Promise<BadCaseInfo> {
  return get<BadCaseInfo>(
    `/projects/${projectId}/bad-cases/${caseId}`,
  );
}

export function updateBadCase(
  projectId: string,
  caseId: string,
  data: {
    status?: BadCaseStatus;
    assignee_id?: string;
    resolution?: string;
  },
): Promise<BadCaseInfo> {
  return put<BadCaseInfo>(
    `/projects/${projectId}/bad-cases/${caseId}`,
    data,
  );
}

export function getBadCaseStats(
  projectId: string,
): Promise<BadCaseStats> {
  return get<BadCaseStats>(
    `/projects/${projectId}/bad-cases/stats`,
  );
}